sql: Auto-commit before DDL in BIND

spilchen opened the pull request:
Several parts of conn_executor_*go already auto-commit transactions before handling DDL commands. However, we were missing the auto-commit for the BIND command. Without it, binding a prepared statement could trigger the auto-commit after the bind, which closes the portal and leads to unknown portal errors.

Epic: none
Release note (bug fix): Fixed a bug when running with autocommit_before_ddl that could cause a runtime error when binding a previously prepared DDL statement.

Fixes #141576
Fixes #141881

cockroach-teamcity commented:
<!-- Reviewable:start -->
This change is [<img src="https://reviewable.io/review_button.svg" height="34" align="absmiddle" alt="Reviewable"/>](https://reviewable.io/reviews/cockroachdb/cockroach/142034)
<!-- Reviewable:end -->

rafiss reviewed (approved):
> *[Reviewable](https://reviewable.io/reviews/cockroachdb/cockroach/142034)* status: <span>:shipit:</span> complete! 0 of 0 LGTMs obtained (waiting on @spilchen)

___
*[`pkg/sql/pgwire/testdata/pgtest/portals_crbugs` line 383 at r1](https://reviewable.io/reviews/cockroachdb/cockroach/142034#-OK1laB90XdCM1lbvZFZ:-OK1laB90XdCM1lbvZF_:b-2kpb2r) ([raw file](https://github.com/cockroachdb/cockroach/blob/a603260bac54970916eba4aa5540b142ea7e501f/pkg/sql/pgwire/testdata/pgtest/portals_crbugs#L383)):*
> ```
> {"Type":"ReadyForQuery","TxStatus":"I"}
> 
> send
> ```

super nit: this portals_crbugs file is intended to test expected differences from PG in how CRDB handles portal exhaustion.

i think a more specific test file for this would be `pkg/sql/pgwire/testdata/pgtest/schema_changes_implicit_txn_with_bind` (though we could rename the file slightly so it's not too specific/verbose).
___
*[`pkg/sql/pgwire/testdata/pgtest/portals_crbugs` line 432 at r1](https://reviewable.io/reviews/cockroachdb/cockroach/142034#-OK1lqNBCpa7Os8u5Fj2:-OK1lqNC2gN0EpVUELE2:b-kvtq3b) ([raw file](https://github.com/cockroachdb/cockroach/blob/a603260bac54970916eba4aa5540b142ea7e501f/pkg/sql/pgwire/testdata/pgtest/portals_crbugs#L432)):*
> ```
> ----
> 
> until
> ```

nit: we could have two different `until` sections: one for postgres and one for cockroachdb. they can be written like this:

this allows the test to pass against postgres (run with `./dev test pkg/sql/pgwire -f=TestPGTest --test-args='-addr localhost:5432 -user user_name'`) as well as cockroachdb.

(this isn't needed if the test is defined in this file, since the file header has `only crdb` at the top, but this will be useful if we move this test to a different file)

```
until noncrdb_only
ReadyForQuery
ReadyForQuery
----
{"Type":"CommandComplete","CommandTag":"BEGIN"}
{"Type":"ReadyForQuery","TxStatus":"T"}
{"Type":"BindComplete"}
{"Type":"CommandComplete","CommandTag":"DROP TABLE"}
{"Type":"ReadyForQuery","TxStatus":"T"}

until crdb_only
ReadyForQuery
ReadyForQuery
----
{"Type":"CommandComplete","CommandTag":"BEGIN"}
{"Type":"ReadyForQuery","TxStatus":"T"}
{"Severity":"NOTICE","SeverityUnlocalized":"NOTICE","Code":"00000","Message":"auto-committing transaction before processing DDL due to autocommit_before_ddl setting","Detail":"","Hint":"","Position":0,"InternalPosition":0,"InternalQuery":"","Where":"","SchemaName":"","TableName":"","ColumnName":"","DataTypeName":"","ConstraintName":"","File":"conn_executor_ddl.go","Line":0,"Routine":"maybeAutoCommitBeforeDDL","UnknownFields":null}
{"Type":"BindComplete"}
{"Type":"CommandComplete","CommandTag":"DROP TABLE"}
{"Type":"ReadyForQuery","TxStatus":"I"}
```


<!-- Sent from Reviewable.io -->

spilchen committed:
sql: Auto-commit before DDL in BIND

Several parts of conn_executor_*go already auto-commit transactions
before handling DDL commands. However, we were missing the auto-commit
for the BIND command. Without it, binding a prepared statement could
trigger the auto-commit after the bind, which closes the portal and
leads to unknown portal errors.

Epic: none
Release note (bug fix): Fixed a bug when running with
autocommit_before_ddl that could cause a runtime error when binding a
previously prepared DDL statement.

Fixes #141576
Fixes #141881

spilchen reviewed (commented):
> *[Reviewable](https://reviewable.io/reviews/cockroachdb/cockroach/142034)* status: <span>:shipit:</span> complete! 0 of 0 LGTMs obtained (waiting on @rafiss and @spilchen)

___
*[`pkg/sql/pgwire/testdata/pgtest/portals_crbugs` line 383 at r1](https://reviewable.io/reviews/cockroachdb/cockroach/142034#-OK1laB90XdCM1lbvZFZ:-OK237oZESVLmMlHhW-b:buww0si) ([raw file](https://github.com/cockroachdb/cockroach/blob/a603260bac54970916eba4aa5540b142ea7e501f/pkg/sql/pgwire/testdata/pgtest/portals_crbugs#L383)):*
<details><summary><i>Previously, rafiss (Rafi Shamim) wrote…</i></summary><blockquote>

super nit: this portals_crbugs file is intended to test expected differences from PG in how CRDB handles portal exhaustion.

i think a more specific test file for this would be `pkg/sql/pgwire/testdata/pgtest/schema_changes_implicit_txn_with_bind` (though we could rename the file slightly so it's not too specific/verbose).
</blockquote></details>

Good point. I moved it to `schema_changes_implicit_txn_with_bind`, and renamed the testcase to `schema_changes_implicit_txn`
___
*[`pkg/sql/pgwire/testdata/pgtest/portals_crbugs` line 432 at r1](https://reviewable.io/reviews/cockroachdb/cockroach/142034#-OK1lqNBCpa7Os8u5Fj2:-OK23I0n9PlN3XOoMdM-:b-4v3ag2) ([raw file](https://github.com/cockroachdb/cockroach/blob/a603260bac54970916eba4aa5540b142ea7e501f/pkg/sql/pgwire/testdata/pgtest/portals_crbugs#L432)):*
<details><summary><i>Previously, rafiss (Rafi Shamim) wrote…</i></summary><blockquote>

nit: we could have two different `until` sections: one for postgres and one for cockroachdb. they can be written like this:

this allows the test to pass against postgres (run with `./dev test pkg/sql/pgwire -f=TestPGTest --test-args='-addr localhost:5432 -user user_name'`) as well as cockroachdb.

(this isn't needed if the test is defined in this file, since the file header has `only crdb` at the top, but this will be useful if we move this test to a different file)

```
until noncrdb_only
ReadyForQuery
ReadyForQuery
----
{"Type":"CommandComplete","CommandTag":"BEGIN"}
{"Type":"ReadyForQuery","TxStatus":"T"}
{"Type":"BindComplete"}
{"Type":"CommandComplete","CommandTag":"DROP TABLE"}
{"Type":"ReadyForQuery","TxStatus":"T"}

until crdb_only
ReadyForQuery
ReadyForQuery
----
{"Type":"CommandComplete","CommandTag":"BEGIN"}
{"Type":"ReadyForQuery","TxStatus":"T"}
{"Severity":"NOTICE","SeverityUnlocalized":"NOTICE","Code":"00000","Message":"auto-committing transaction before processing DDL due to autocommit_before_ddl setting","Detail":"","Hint":"","Position":0,"InternalPosition":0,"InternalQuery":"","Where":"","SchemaName":"","TableName":"","ColumnName":"","DataTypeName":"","ConstraintName":"","File":"conn_executor_ddl.go","Line":0,"Routine":"maybeAutoCommitBeforeDDL","UnknownFields":null}
{"Type":"BindComplete"}
{"Type":"CommandComplete","CommandTag":"DROP TABLE"}
{"Type":"ReadyForQuery","TxStatus":"I"}
```
</blockquote></details>

I didn't know you could do this with those tests. Since I moved the test that isn't crdb_only, I added the separate until sections and confirmed it ran on a postgres instance.


<!-- Sent from Reviewable.io -->

spilchen commented:
TFTR!

bors r+

spilchen commented:
I will backport this in https://github.com/cockroachdb/cockroach/pull/141851 and https://github.com/cockroachdb/cockroach/pull/141987
